mmdet/structures/bbox/horizontal_bbox.py

```diff
@@ -0,0 +1,324 @@
+
+    def rotate_(self, center: Tuple[float, float], angle: float) -> None:
+        """Inplace rotate all boxes.
+
+        Args:
+            center (Tuple[float, float]): Rotation origin.
+            angle (float): Rotation angle represented in degrees.
+        """
+        bboxes = self.tensor
+        rotation_matrix = bboxes.new_tensor(
+            cv2.getRotationMatrix2D(center, angle, 1))
+
+        corners = self.hbbox2corner(bboxes)
+        corners = torch.cat(
+            [corners, corners.new_ones(*corners.shape[:-1], 1)], dim=-1)
+        corners_T = torch.transpose(corners, -1, -2)
+        corners_T = torch.matmul(rotation_matrix, corners_T)
+        corners = torch.transpose(corners_T, -1, -2)
+        self.tensor = self.corner2hbbox(corners)
+
+    def project_(self, homography_matrix: Union[Tensor, np.ndarray]) -> None:
+        """Inplace geometric transformation for bbox.
+
+        Args:
+            homography_matrix (Tensor or np.ndarray]):
+                Shape (3, 3) for geometric transformation.
+        """
+        bboxes = self.tensor
+        if isinstance(homography_matrix, np.ndarray):
+            homography_matrix = bboxes.new_tensor(homography_matrix)
+        corners = self.hbbox2corner(bboxes)
+        corners = torch.cat(
+            [corners, corners.new_ones(*corners.shape[:-1], 1)], dim=-1)
+        corners_T = torch.transpose(corners, -1, -2)
+        corners_T = torch.matmul(homography_matrix, corners_T)
+        corners = torch.transpose(corners_T, -1, -2)
+        # Convert to homogeneous coordinates by normalization
+        corners = corners[..., :2] / corners[..., 2:3]
+        self.tensor = self.corner2hbbox(corners)
+
+    @staticmethod
+    def hbbox2corner(bboxes: Tensor) -> Tensor:
+        """Convert bbox coordinates from (x1, y1, x2, y2) to corners ((x1, y1),
+        (x2, y1), (x1, y2), (x2, y2)).
+
+        Args:
+            bboxes (Tensor): Horizontal box tensor with shape of (..., 4).
+
+        Returns:
+            Tensor: Corner tensor with shape of (..., 4, 2).
+        """
+        x1, y1, x2, y2 = torch.split(bboxes, 1, dim=-1)
+        corners = torch.cat([x1, y1, x2, y1, x1, y2, x2, y2], dim=-1)
+        return corners.reshape(*corners.shape[:-1], 4, 2)
+
+    @staticmethod
+    def corner2hbbox(corners: Tensor) -> Tensor:
+        """Convert bbox coordinates from corners ((x1, y1), (x2, y1), (x1, y2),
+        (x2, y2)) to (x1, y1, x2, y2).
+
+        Args:
+            corners (Tensor): Corner tensor with shape of (..., 4, 2).
+
+        Returns:
+            Tensor: Horizontal box tensor with shape of (..., 4).
+        """
+        min_xy = corners.min(dim=-2)[0]
+        max_xy = corners.max(dim=-2)[0]
+        return torch.cat([min_xy, max_xy], dim=-1)
+
+    def rescale_(self,
+                 scale_factor: Tuple[float, float],
+                 mapping_back=False) -> None:
+        """Inplace rescale boxes w.r.t. rescale_factor.
```

Comment:
I consider it will be convenient in `_bbox_post_process`. We can directly input img_meta['scale_factor'] and don't need to calculate the mapping back scales.

Reply:
`bboxes / scale_factor` is the opposite of the normal way, so you can to describe this in docstring.

Reply:
Finally, I decide to delete this argument.